### PR TITLE
Add a test, Fixed to assert either IndexError or Exception message

### PR DESCRIPTION
**Pull request by positive235**
**1. Add a variable ‘YOUR_EXCEPTION_MESSAGE’ which can be assigned to Exception message the user assigned. Default exception message is set to: “Index out of bounds”**

**2. Add test_add_link_before_out_of_bounds_index_n():**
- tries to insert D at index 3 when having 3 elements, verifies either IndexError, or Exception and its message is raised.

**3. Fixed tests in order to assert either IndexError or, Exception and its message**

**4. Changes**
- test_add_link_before_empty_list() - Changed to compare with string 
- test_add_link_before() - Changed to compare with string

### CODE REVIEW

File: linked_list_unit_test.py [hunk 1]
```diff
@@ -29,7 +29,16 @@ def test_add_link_before(self):
...
-
+    
+    # Tries to insert D at index 3, verifies an IndexError is raised. 
+    # (If list if not empty and have, say, N elements, valid indices for both methods are 0 ... N-1, inclusive.)
+    # CHECK the related piazza post: https://piazza.com/class/k8aktyvmw804ns?cid=141
+    def test_add_link_before_out_of_bounds_index_n(self):
+        self.ll.add_link_before("A", 0)
+        self.ll.add_link_before("B", 0)
+        self.ll.add_link_before("C", 1)
+        self.ll.add_link_before("D", 3) # -> Exception: Index out of bounds
```
Comment (gkochera) [inline linked_list_unit_test.py]: This needs to be wrapped in a with self.assertRaises(IndexError) + try catch, in order to verify the operation actually raises the exception.

**Review by gkochera — CHANGES_REQUESTED** [review]
I like all of your changes. Can you just fix the one thing on test_add_link_before_out_of_bounds_index_n(self): This function is also very similar to test_add_link_before_fails_at_link_before_tail(self)... verify that its needed.

File: linked_list_unit_test.py (hunk 1 above)
Comment (positive235) [reply linked_list_unit_test.py]: The function test_add_link_before_out_of_bounds_index_n has already changed with another codes in order to verify IndexError or Exception and its message. Refers to "Files changed":
```
    # Tries to insert D at index 3, verifies an IndexError is raised. 
    # (If list if not empty and have, say, N elements, valid indices for both methods are 0 ... N-1, inclusive.)
    # CHECK the related piazza post: https://piazza.com/class/k8aktyvmw804ns?cid=141
    def test_add_link_before_out_of_bounds_index_n(self):
        self.ll.add_link_before("A", 0)
        self.ll.add_link_before("B", 0)
        self.ll.add_link_before("C", 1)

        try:
            with self.assertRaises(IndexError):
                self.ll.add_link_before("D", 3)
        except Exception as msg:
             self.assertEqual(msg.args[0], YOUR_EXCEPTION_MESSAGE)
```

Comment (positive235) [reply linked_list_unit_test.py]: And the function test_add_link_before_fails_at_link_before_tail(self) is not the same as test_add_link_before_out_of_bounds_index_n(self). The former verifies the add_link_before function, but the latter verifies add_front function.

Comment (positive235) [reply linked_list_unit_test.py]: After having dinner (in 2 hours? my husband just cooked so I can’t put off eating, sorry) I’m gonna fix another functions which requires verifications in order to check IndexError or Exception message simultaneously.

Comment (positive235) [reply linked_list_unit_test.py]: Thanks for reviewing! The function test_add_link_before_out_of_bounds_index_n had already changed with another codes in order to verify IndexError or Exception and its message. 
```
    # Tries to insert D at index 3, verifies an IndexError is raised. 
    # (If list if not empty and have, say, N elements, valid indices for both methods are 0 ... N-1, inclusive.)
    # CHECK the related piazza post: https://piazza.com/class/k8aktyvmw804ns?cid=141
    def test_add_link_before_out_of_bounds_index_n(self):
        self.ll.add_link_before("A", 0)
        self.ll.add_link_before("B", 0)
        self.ll.add_link_before("C", 1)

        try:
            with self.assertRaises(IndexError):
                self.ll.add_link_before("D", 3)
        except Exception as msg:
             self.assertEqual(msg.args[0], YOUR_EXCEPTION_MESSAGE)
```

+ Update: Fixed another verification tests in order to check IndexError or Exception messages at the same time

**Pull request merged**